xcm/src/v1/multi_location.rs

```diff
@@ -0,0 +1,867 @@
+	/// assert_eq!(
+	///     m.match_and_split(&MultiLocation::new(1, X1(PalletInstance(3))).unwrap()),
+	///     Some(&OnlyChild),
+	/// );
+	/// assert_eq!(m.match_and_split(&MultiLocation::new(1, Null).unwrap()), None);
+	/// # }
+	/// ```
+	pub fn match_and_split(&self, prefix: &MultiLocation) -> Option<&Junction> {
+		if self.parents != prefix.parents {
+			return None
+		}
+		self.junctions.match_and_split(&prefix.junctions)
+	}
+
+	/// Mutate `self` so that it is suffixed with `suffix`. The correct normalized form is returned,
+	/// removing any internal [Non-Parent, `Parent`]  combinations.
+	///
+	/// Does not modify `self` and returns `Err` with `suffix` in case of overflow.
+	///
+	/// # Example
+	/// ```rust
+	/// # use xcm::v1::{Junctions::*, Junction::*, MultiLocation};
+	/// # fn main() {
+	/// let mut m = MultiLocation::new(1, X2(Parachain(21), OnlyChild)).unwrap();
+	/// assert_eq!(m.append_with(MultiLocation::new(1, X1(PalletInstance(3))).unwrap()), Ok(()));
+	/// assert_eq!(m, MultiLocation::new(1, X2(Parachain(21), PalletInstance(3))).unwrap());
+	/// # }
+	/// ```
+	pub fn append_with(&mut self, suffix: MultiLocation) -> Result<(), MultiLocation> {
+		let mut prefix = suffix;
+		core::mem::swap(self, &mut prefix);
+		match self.prepend_with(prefix) {
+			Ok(()) => Ok(()),
+			Err(prefix) => {
+				let mut suffix = prefix;
+				core::mem::swap(self, &mut suffix);
+				Err(suffix)
+			}
+		}
+	}
+
+	/// Mutate `self` so that it is prefixed with `prefix`.
+	///
+	/// Does not modify `self` and returns `Err` with `prefix` in case of overflow.
+	///
+	/// # Example
+	/// ```rust
+	/// # use xcm::v1::{Junctions::*, Junction::*, MultiLocation};
+	/// # fn main() {
+	/// let mut m = MultiLocation::new(2, X1(PalletInstance(3))).unwrap();
+	/// assert_eq!(m.prepend_with(MultiLocation::new(1, X2(Parachain(21), OnlyChild)).unwrap()), Ok(()));
+	/// assert_eq!(m, MultiLocation::new(1, X1(PalletInstance(3))).unwrap());
+	/// # }
+	/// ```
+	pub fn prepend_with(&mut self, mut prefix: MultiLocation) -> Result<(), MultiLocation> {
+		let self_parents = self.parent_count();
+		let prepend_len = (self_parents as isize - prefix.junctions.len() as isize).abs() as usize;
+		if self.junctions.len() + prefix.parent_count() + prepend_len > MAX_MULTILOCATION_LENGTH {
+			return Err(prefix)
+		}
+
+		let mut final_parent_count = prefix.parents;
+		for _ in 0..self_parents {
+			if prefix.take_last().is_none() {
+				// If true, this means self parent count is greater than prefix junctions length;
+				// add the resulting self parent count to final_parent_count
+				final_parent_count += self.parents;
+				break
+			}
+			self.pop_parent();
+		}
+
+		self.parents = final_parent_count;
+		for j in prefix.junctions.into_iter_rev() {
+			self.push_front_non_parent(j).expect(
+				"self junctions len + prefix parent count + prepend len is less than max length; qed"
+			);
+		}
+		Ok(())
+	}
+}
+
+impl From<Junctions> for MultiLocation {
```

Comment:
There should alss be an `impl TryFrom<MultiLocation> for Junctions` which fails if `parents > 0`.